src/halmos/sevm.py

```diff
@@ -2302,24 +2342,21 @@ def call_unknown() -> None:
 
             last_idx = len(ret_lst) - 1
             for idx, ret_ in enumerate(ret_lst):
+                if not isinstance(ret_, ByteVec):
+                    raise HalmosException(f"Invalid return value: {ret_}")
+
                 new_ex = (
                     self.create_branch(ex, BoolVal(True), ex.pc)
                     if idx < last_idx
                     else ex
                 )
 
                 # TODO: refactor this return memory setting to be shared with call_known()
-                # store return value
+                # copy return data to memory
                 effective_ret_size = min(ret_size, len(ret_))
                 if effective_ret_size > 0:
-                    end_loc = ret_loc + effective_ret_size
-                    if end_loc > MAX_MEMORY_SIZE:
-                        raise HalmosException("returned data exceeds MAX_MEMORY_SIZE")
-
-                    new_ex.st.memory.set_slice(ret_loc, end_loc, ret_)
-
-                if not isinstance(ret_, ByteVec):
-                    raise HalmosException(f"Invalid return value: {ret_}")
+                    data = ret_.slice(0, effective_ret_size)
```

Comment:
if i understand correctly, this needs the full scan even if `effective_ret_size == len(ret_)`. could you make sure slice() returns immediately in such case?

Reply:
good point, let me look into it

Reply:
in 48424ca

Reply:
https://github.com/a16z/halmos/commit/48424caf7a65ef1c69bd9d3327fc3d381eeda5c8 looks good. but it addressed the same issue in a different location, leaving this one unaddressed. alternatively, you might consider improving the bytevec.slice(0, size) function to immediately return when size is the full length?

Reply:
> you might consider improving the bytevec.slice(0, size) function to immediately return when size is the full length?

that may be a little tricky to handle properly, because bytevecs are mutable. So in principle we need to return a copy, but in this case we can skip the copy because we just blit it in memory and we know we're not going to modify the returndata any further

Reply:
I'll handle it at this callsite and we can think about reducing copies further or introducing immutable bytevecs in a separate PR

Reply:
sorry for the confusion, I refactored both places to call the same function: ebc6da6